Add recipe for easy-tabulated-list.

jcs090218 opened the pull request:
### Brief summary of what the package does

Simplify usage for `tabulated-list'.

### Direct link to the package repository

https://github.com/jcs-elpa/easy-tabulated-list

### Your association with the package

The maintainer.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

alphapapa commented:
How does this package compare to these two packages which seem to serve the same purpose?

* `navigel`
* `tblui`

jcs090218 commented:
> How does this package compare to these two packages which seem to serve the same purpose?

Okay, I have look into these packages. `tblui` serves similar but different purpose, and it is almost something I am looking for. The problem is that both packages depends on `tablist` and I don't really uses this package. This package only focuses on the basic implementation of the `tabulated-list`. I guess this is more of the personal references?

riscy commented:
> tblui serves similar but different purpose, and it is almost something I am looking for.  The problem is that both packages depends on tablist and I don't really uses this package.

Hi!  The code looks good, but can you explain more about how the purpose differs?  I'm also not entirely convinced that it is a problem that those packages have a dependency you don't use.  A detailed better explanation would be much appreciated, since it can save a lot of reviewer time.

jcs090218 commented:
> Hi! The code looks good, but can you explain more about how the purpose differs?

I think `tablui` has many feature base on another extension like, `magit-popup`, and mainly `tablist`. This is a strip version of `tablui` that only focuses on Emacs built-in `tabulated-list`. This package serves only to users that want to build `tabulated-list` (the table) easily and no other purpose. I assumed if the user wants more then they should go ahead and use `tablui`.

riscy commented:
Thanks for that.  You left out navigel -- how does its purpose differ?  Based on the READMEs alone, I had trouble teasing out the difference.  tblui is a bit long in the tooth, but both it and navigel are well documented with _complete_ examples, which I think is something you should strive for in an "easy-<xyz>" wrapper package.  (The examples in your README had values stubbed out that I realized I'd need to read the documentation for `tabulated-list` to fill in.)

Separately, while I agree with your bias toward fewer dependencies, `tablist` looks well maintained and is of apparently high quality.  Could you please explain why it is important to you to avoid this dependency?

And lastly, I wanted to reach out here and apologize for how many of your open PR's are sitting unreviewed.  I am trying to divide my time across authors rather than pull requests, and you have the most pull requests.

jcs090218 commented:
> Separately, while I agree with your bias toward fewer dependencies, tablist looks well maintained and is of apparently high quality. Could you please explain why it is important to you to avoid this dependency?

I'm not trying to avoid this dependency. I think `tablist` has a lot more feature to extend `tabulated-list`; and I assumed if user only wants to extend `tabulated-list` and not `tablist` will want something simpler and direct?

> (The examples in your README had values stubbed out that I realized I'd need to read the documentation for tabulated-list to fill in.)

This make sense because this package is built on top of `tabulated-list` and not `tablist`. I think I would call this package `easy-tablist` if I was trying to extend something on top of `tablist` and not `tabulated-list`.

> You left out navigel -- how does its purpose differ?

I don't really know much about `navigel`. I have read through the README.md, examples, and video [here](https://www.youtube.com/watch?v=rTxFnMiAtQo&ab_channel=Emacsconferencesandhangouts), but I don't feel like it is similar to `easy-tabulated-list` or `tablui`? 😕 Am I missing something?

I feel like navigel has a different user than a wrapper package?

> Based on the READMEs alone, I had trouble teasing out the difference. tblui is a bit long in the tooth, but both it and navigel are well documented with complete examples, which I think is something you should strive for in an "easy-" wrapper package.

I consider my package as a tiny library that focus on building table on top of `tabulated-list` and nothing else. The user of this package is simple, they only want to build something on top of `tabulated-list` and not `tablist`. Now I can imagine why `tablui` is made; and it make sense. 😆 If the user want something from `tablist` then they should use `navigel` or `tablui` and not `easy-tabulated-list`.

Of course, you can use `navigel` or `tablui` to achieve the same thing. But you would just have an un-necessary layer inside your tool, and I don't think this is a elegant code. 😕 

And yes, I do need to improve my documentation here. I kind of assumed people using this library knows `tabulated-list`; not `tablist`. 😅 

> And lastly, I wanted to reach out here and apologize for how many of your open PR's are sitting unreviewed. I am trying to divide my time across authors rather than pull requests, and you have the most pull requests.

No worries! You are doing a great job reviewing packages and it cost a lost of time. I very appreciate your patience and your effort in Emacs community! Thank you!

riscy commented:
> And yes, I do need to improve my documentation here.

Thanks -- I should have been more direct and suggest better documentation and examples are things I feel are missing for this package to feel complete.  Feel free to ping when ready with those.

jcs090218 commented:
Okay, I have added more details in the README file. I have also added a example file to the repository, see [here](https://github.com/jcs-elpa/easy-tabulated-list/blob/master/examples/test-table.el).

riscy commented:
Okay, thanks for the example. That helps clarify things.

- Since this is meant to be an "easy" interface, you shouldn't inject your personal preferences for the tabulated-list defaults, in particular by making the default padding 1 instead of 0
- As a rule of thumb, don't hook into the private implementation details of another package. In this case your fake-header relies on `tabulated-list--header-string`.  If you read the docs, that private string is meant to be "populated by tabulated-list-init-header", in particular when `tabulated-list-use-header-line` is nil
- I mentioned this in another one of your PRs, but aim to go a bit further than just copying your summary line into your Commentary.  "This package provides simplified interfaces for transforming lists and vectors into a format supported by tabulated-list (see function xyz), and for creating buffers which display tabulated lists (see function abc)" or something like that.

Minor stuff:
- In your example code, `test-table-data` is set to nil, which makes for a bland (and initially confusing) example
- Throughout, when you use the verb "form" you probably mean "format"
- "without a valid ... is set" -> "without a valid ... set"

jcs090218 commented:
> As a rule of thumb, don't hook into the private implementation details of another package. In this case your fake-header relies on tabulated-list--header-string. If you read the docs, that private string is meant to be "populated by tabulated-list-init-header", in particular when tabulated-list-use-header-line is nil

What do you mean? So I shouldn't touch `tabulated-list--header-string` variable. Is there a way to get around this? 😕 

---

Other issues are fixed. Thank you!

riscy commented:
I've carefully double-checked my understanding of tabulated-list to ensure I'm not coming off as unreasonable here.

I want to help ensure MELPA is a welcoming place, especially for new authors, but you are not new and you should know better.  If you don't understand what I mean, then it's because you're not putting in the effort to understand tabulated-list implementation details.  I clearly pointed you to the function and variable you need to read about.  Replying "what do you mean?" isn't even meeting me halfway.

It's okay to not know.  We all have limited attention.  But this is a sign you should slow down and spend more time with your packages before putting them in front of someone else and expecting a heartfelt review.  Please.

You can tell I'm at the end of my rope, but I'll still try to be helpful.  The local variable `tabulated-list-use-header-line` determines whether `header-line-format` is going to be used to display the column names.  If and only if it is false, then the (private) variable `tabulated-list--header-string` is populated with the column names.  There is no public interface for setting this private variable except implicitly through `tabulated-list-use-header-line`, and a wrapper should respect that.

jcs090218 commented:
Got it! Thank you! Close this issue now!
